kong/plugins/ai-proxy/handler.lua

```diff
@@ -0,0 +1,183 @@
+local _M = {}
+
+-- imports
+local ai_shared = require("kong.plugins.ai-proxy.drivers.shared")
+local fmt = string.format
+local kong_utils = require("kong.tools.utils")
+local kong_meta = require "kong.meta"
+--
+
+_M.PRIORITY = 769
+_M.VERSION = kong_meta.version
+
+local INTERNAL_SERVER_ERROR_TEMPLATE = '{"error":{"message":"%s"}}'
+
+local function do_bad_request(msg)
+  kong.log.warn(msg)
+  kong.response.exit(400, { message = msg, error = true })
+end
+
+local function do_internal_server_error(msg)
+  kong.log.warn(msg)
+  kong.response.exit(500, { message = msg, error = true })
+end
+
+local function apply_transformations(route_type, request_table, transformation_instructions)
+  -- this is mostly a placeholder function for now;
+  -- it just receives a replacement "template" from a previous
+  -- AI-plugin, and applies it over the top of the existing request body
+  local this = {}
+   
+  -- use template engine
+  if route_type == "llm/v1/completions" then
+    kong.log.debug("template is llm/v1/completions type")
+    this = {
+      prompt = transformation_instructions[1],
+    }
+  elseif route_type == "llm/v1/chat" then
+    kong.log.debug("template is llm/v1/chat type")
+    this = {
+      messages = {
+        [1] = {
+          role = "system",
+          content = "You are a helpful assistant"
+        },
+        [2] = {
+          role = "user",
+          content = transformation_instructions[1]
+        }
+      }
+    }
+  end
+
+  return this
+end
+
+function _M:header_filter(conf)
+  kong.log.debug("IN: ai-proxy/header_filter")
+
+  -- clear shared restricted headers
+  for i, v in ipairs(ai_shared.clear_response_headers.shared) do
+    kong.response.clear_header(v)
+  end
+
+  -- only act on 200 in first release - pass the unmodifed response all the way through if any failure
+  if kong.response.get_status() == 200 then
+    local content_type = kong.response.get_header("content-type") or "application/json"
+    if content_type then
+      local ai_driver = require("kong.plugins.ai-proxy.drivers." .. conf.model.provider)
+      local route_type = conf.route_type
+      
+      local response_body = kong.service.response.get_raw_body()
```

Comment:
it will make more sense to transforma and use the response body both in `body_filter` phase. it will be confusing to set it in header_filter and use in body_filter.

Reply:
@fffonion We can't do that, because we can't change the status code (on parse error) in the body_filter phase.

Reply:
@fffonion If there is a way to both GET and SET body, status code, and headers, all in one function? I can use this?

Otherwise I am a bit stuck here, seems like limitation of nginx, even with request_buffering on.